pyscriptjs/src/main.ts

```diff
@@ -61,46 +92,117 @@ class PyScriptApp {
+    // lifecycle (3)
+    showLoader() {
+        // add loader to the page body
+        logger.info('add py-loader');
+        customElements.define('py-loader', PyLoader);
+        this.loader = <PyLoader>document.createElement('py-loader');
+        document.body.append(this.loader);
     }
 
-    loadPackages = async () => {
+    // lifecycle (4)
+    loadRuntime() {
+        logger.info('Initializing runtime');
+        if (this.config.runtimes.length == 0) {
+            showError("Fatal error: config.runtimes is empty");
+            return;
+        }
+
+        if (this.config.runtimes.length > 1) {
+            showError("Multiple runtimes are not supported yet. " +
+                      "Only the first will be used");
+        }
+        const runtime_cfg = this.config.runtimes[0];
+        const runtime: Runtime = new PyodideRuntime(this.config, runtime_cfg.src,
+                                                    runtime_cfg.name, runtime_cfg.lang);
+        this.loader.log(`Downloading ${runtime_cfg.name}...`);
+        const script = document.createElement('script'); // create a script DOM node
+        script.src = runtime.src;
+        script.addEventListener('load', () => {
+            void this.afterRuntimeLoad(runtime);
+        });
+        document.head.appendChild(script);
+    }
+
+    // lifecycle (5)
+    // See the overview comment above for an explanation of how we jump from
+    // point (4) to point (5).
+    //
+    // Invariant: this.config and this.loader are set and available.
+    async afterRuntimeLoad(runtime: Runtime): Promise<void> {
+        // XXX what is the JS/TS standard way of doing asserts?
+        console.assert(this.config !== undefined);
+        console.assert(this.loader !== undefined);
+
+        this.loader.log('Python startup...');
+        await runtime.loadInterpreter();
+        runtimeLoaded.set(runtime);
+        this.loader.log('Python ready!');
+
+        // eslint-disable-next-line
+        runtime.globals.set('pyscript_loader', this.loader);
+
+        this.loader.log('Setting up virtual environment...');
+        await this.setupVirtualEnv(runtime);
+        await mountElements(runtime);
+
+        this.loader.log('Executing <py-script> tags...');
+        this.executeScripts(runtime);
+
+        this.loader.log('Initializing web components...');
+        // lifecycle (8)
+        createCustomElements();
+
+        if (runtime.config.autoclose_loader) {
+            this.loader.close();
+        }
+        await initHandlers(runtime);
+
+        // NOTE: runtime message is used by integration tests to know that
+        // pyscript initialization has complete. If you change it, you need to
+        // change it also in tests/integration/support.py
+        logger.info('PyScript page fully initialized');
+    }
+
+
+    // lifecycle (6)
+    async setupVirtualEnv(runtime: Runtime): Promise<void> {
+        // XXX: maybe the following calls could be parallelized, instead of
```

Comment:
parallel loading
```js
await Promise.all([
    runtime.installPackage(this.config.packages),
    this.fetchPaths(runtime)
])
```

Reply:
thanks! This looks like a very useful idiom, I'll keep it in mind.
But I don't want to fix it in this PR: I prefer to keep the logic changes at minimum, so that if we break something without noticing, it will be easier to `git bisect`.
Basically, I wrote these `XXX` as a mean to say "we know this is suboptimal and we should solve it, but not now".